Zilliqa/contracts/xsgd_contract.scilla

```diff
@@ -574,41 +576,39 @@ transition Transfer(to : ByStr20, amount : Uint128, initiator : ByStr20, to_bal
   send msgs
 end
 
-transition TransferFrom (from : ByStr20, to : ByStr20, amount : Uint128, initiator : ByStr20, to_bal : Uint128, from_bal : Uint128)
+transition TransferFrom (from : ByStr20, to : ByStr20, amount : Uint128, initiator : ByStr20, to_bal : Uint128, from_bal : Uint128, spender_allowance : Uint128)
   (* Conditional checks *)  
   isProxy _sender;
   isNotPaused;
   isNotBlacklisted initiator;
   isNotBlacklisted to;
   isNotBlacklisted from;
-  isNotSelf from to;
+  isNotSender from to;
   isNotNull to;
-  
-  (* Checks if amount to transfer <= allowance of spender *)
-  option_spender_allowance <- allowed[from][initiator];
-  spender_allowance = deconstruct_option_uint128 option_spender_allowance;
 
-  (* Checks if amount to transfer <= min(balance of from, spender's allowance) *)
-  min = min_int from_bal spender_allowance;
-  withinLimits min amount;
+  (* There is an optimisation here that I removed because it results in not very descriptive errors. *)
+  (* The optimisation involves getting the min(bal, allowance) and comparing the result with amount. *)
+  isSufficientFunds from_bal amount;
+  isSufficientAllowance spender_allowance amount;
 
   (* Calculates updated balances of to and from *)
   new_from_bal = builtin sub from_bal amount;
   new_to_bal = builtin add to_bal amount;
   new_spender_allowance = builtin sub spender_allowance amount;
-  allowed[from][initiator] := new_spender_allowance;
   
   (* Emits event *)
   e = {_eventname : "TransferFromSuccess"; initiator : initiator; sender : from; recipient : to; amount : amount};
   event e;
   
-  (* Create and sends the transaction to be called on the proxy contract since balances are stored on the proxy, not the implementation *)
-  msg_to_proxy = {_tag : "TransferFromCallBack"; _recipient : _sender; _amount : zero;
+  (* Create and sends the transaction to be called on the proxy contract since balances and allowances are stored on the proxy, not the implementation *)
+  msg_to_proxy_balances = {_tag : "TransferFromCallBack"; _recipient : _sender; _amount : zero;
           from : from; to : to; new_from_bal : new_from_bal; new_to_bal : new_to_bal };
+  msg_to_proxy_allowance = {_tag : "AllowanceCallBack"; _recipient : _sender; _amount : zero;
+          initiator : from; spender : initiator; new_allowance : new_spender_allowance};
   msg_to_sender = {_tag : "TransferFromSuccessCallBack"; _recipient : initiator; _amount : zero;
           initiator : initiator; sender : from; recipient : to; amount : amount };
   msg_to_recipient = {_tag : "RecipientAcceptTransferFrom"; _recipient : to; _amount : zero;
           initiator : initiator; sender : from; recipient : to; amount : amount};
-  msgs = three_msgs msg_to_proxy msg_to_sender msg_to_recipient;
+  msgs = four_msgs msg_to_proxy_balances msg_to_proxy_allowance msg_to_sender msg_to_recipient;
```

Comment:
The other time you mentioned still not being very sure about the order of the callbacks, has that been settled?

Reply:
you mean the order in which the callbacks are handled? Yeah, I asked @evesnow91 a while back re. this. Basically it's acts like a stack so the last one in is the first one out.